### ISSUE TITLE
cucumber reorder scenario execution

### ISSUE THREAD

**Issue by akostadinov** [cucumber/cucumber#789]
cucumber 1.3 executed scenarios in the order specified on the command line. Now it does some kind of sorting and no matter of command line order, it runs them in the same order. Originally I reported the issue in #778 along with other related issues.

```
$ cucumber --version
2.0.0.beta.5
$ cucumber ./features/3scenarios.feature:9:6
Feature: ASD

  Scenario: gah2     # ./features/3scenarios.feature:6
    Given I love you # features/step_definitions/gah.rb:1
      hey

  Scenario: gah3     # ./features/3scenarios.feature:9
    Given I love you # features/step_definitions/gah.rb:1
      hey

2 scenarios (2 passed)
2 steps (2 passed)
0m0.042s
$ cucumber ./features/3scenarios.feature:6:9
Feature: ASD

  Scenario: gah2     # ./features/3scenarios.feature:6
    Given I love you # features/step_definitions/gah.rb:1
      hey

  Scenario: gah3     # ./features/3scenarios.feature:9
    Given I love you # features/step_definitions/gah.rb:1
      hey

2 scenarios (2 passed)
2 steps (2 passed)
0m0.041s

$ cucumber ./features/3scenarios.feature:6 ./features/3scenarios.feature:9
Feature: ASD

  Scenario: gah2     # ./features/3scenarios.feature:6
    Given I love you # features/step_definitions/gah.rb:1
      hey

  Scenario: gah3     # ./features/3scenarios.feature:9
    Given I love you # features/step_definitions/gah.rb:1
      hey

Feature: ASD

  Scenario: gah2     # ./features/3scenarios.feature:6
    Given I love you # features/step_definitions/gah.rb:1
      hey

  Scenario: gah3     # ./features/3scenarios.feature:9
    Given I love you # features/step_definitions/gah.rb:1
      hey

4 scenarios (4 passed)
4 steps (4 passed)
0m0.045s
```

This is the bare minimum test project: https://github.com/akostadinov/testcuke

btw I still see the duplicate execution issue as visible in the last cucumber run above as I reported in #778. I'm running cucumber 2.0.0.beta5

**Comment by jdks** [comment]
> btw I still see the duplicate execution issue as visible in the last cucumber run above as I reported in > #778. I'm running cucumber 2.0.0.beta5

@akostadinov, thanks for the detailed update regarding duplicate execution. I think the fix for that issue is not currently included in `2.0.0.beta.5` and may go out in a future release (the core team would be able to confirm).

FYI I'm able to get the following results by pointing to a more recent version of the cucumber gem in a [fork](https://github.com/jdks/testcuke) of your example.

```
$ bundle exec cucumber --format progress features/3scenarios.feature:6 features/3scenarios.feature:9

hey
.
hey
.

2 scenarios (2 passed)
2 steps (2 passed)
0m0.018s
```

**Comment by akostadinov** [comment]
@jdks, ok, so duplicates are removed, only reordering still occurring, thanks

**Comment by brasmusson** [comment]
@akostadinov @jdks Yes, as [History.md](https://github.com/cucumber/cucumber/blob/master/History.md) says, #786, which fixes the duplication issue, is on the master branch and will be included in the next release.

**Comment by mattwynne** [comment]
Hi @akostadinov, I can confirm the behaviour is as you've described. I'll explain why it works like this at the moment.

When we run your features, we send each feature file through a compiler to generate test cases. Each scenario becomes a test case, as does each row in any examples tables you have. The `file:line:line` argument you pass on the command-line is a filter. Each test case is passed, in turn, through the filters you've specified (you might also specify other filters, like tags). Once the test case makes it through the filters you've specified, we start executing it.

So the ordering you're experiencing simply comes from the order of the scenarios in the feature file.

Can you explain why it's bothering you? Is there a reason you wan to be able to run the scenarios in a particular order?

**Comment by akostadinov** [comment]
@mattwynne, the reason I want the original order is so that multiple processes going over the same set of test cases has smaller chance of executing the same test at the same time. We use an external test tracking system to avoid duplicate runs but it's not ACID and reordering really helps. Actually if cucumber has an option to randomize execution, that can also help. Or if I can do it programatically from env.rb that can also work for our purposes.

A less significant reason I have is that sometimes I want to run a number of scenarios where I want to run the more riskier first so I know about issues earlier. It is more convenient to prepare a list of scenarios than running one after the other manually. 

I'm curious though, how duplicates happen to appear as in #778 given your explanation of how scenarios are filtered. Because doing `.uniq` in #786 looks like just a workaround to the real problem.

I don't know how you're doing the filtering. If there's any feasible way all individual filters to have their own associated bucket, then when a filter matches it could insert the scenario into it's respective bucket and finally `flatten` the buckets in to a single one in the filter order. I'm guessing though things are more complicated than that.

**Comment by mattwynne** [comment]
Yeah I can see how this happens now.

The `.uniq` call [here](https://github.com/cucumber/cucumber/blob/master/lib/cucumber/file_specs.rb#L19) is used to generate the list of files to be parsed. Parsed files are immediately fed into the filter/execution chain. So when that list contained dupes, we were parsing (and therefore executing) the duplicate files twice.

**Comment by mattwynne** [comment]
The [locations filter](https://github.com/cucumber/cucumber-ruby-core/blob/master/lib/cucumber/core/test/filters/locations_filter.rb) we currently use only filters out scenarios, it does not sort them.

What we need to solve this issue is to create another filter that sorts test cases, based on the list of locations. My preference is to have this separate to the existing `Core::Test::LocationsFilter` just to keep everything modular and single-responsibility.

Anyone want to have a crack at building that sorter?

**Comment by akostadinov** [comment]
Where should that sorter be hooked? The location filter appears to just compare a case to the locations and if it is included, pass it down.
To sort things on the other hand, one needs to pile the scenarios and then sort them out, then finally pass them down. I imagine this can be emulated in a filter but how would that filter know when the last scenario has been reached and pass down the scenarios to be executed?
I think that easiest would be to reuse the locations filter as it does locations matching anyway and having a separate filter would require matching the scenarios to locations again.. double work. But still the question remains - how does filter know when last scenario has been reached to pass down the sorted scenarios.
At least this is my ignorant view on the issue.

**Comment by mattwynne** [comment]
Should be fixed by b43e5d59958eb623ab5370be676ea414f61e6f06

**Comment by mattwynne** [comment]
This will be included in the next release. I'm going to close it but please do open another ticket if it's not working when we release rc4
